### ISSUE TITLE
companion object custom default serializer pattern fails to build for mingw

### ISSUE THREAD

**Issue by benasher44** [Kotlin/kotlinx.serialization#566]
**Describe the bug**
Using the `companion object` custom default serializer pattern, code fails to build for mingw64. Here's the output:
```
* Compiler version info: Konan: 1.3.50 / Kotlin: 1.3.50
 * Output kind: DYNAMIC

e: java.lang.IllegalStateException: no function <init> in org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl@568b2d3b
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGenerator.llvmFunction(CodeGenerator.kt:23)
	at org.jetbrains.kotlin.backend.konan.ExportedElement.generateCAdapter(CAdapterGenerator.kt:220)
	at org.jetbrains.kotlin.backend.konan.ExportedElementScope.generateCAdapters(CAdapterGenerator.kt:170)
	at org.jetbrains.kotlin.backend.konan.ExportedElementScope.generateCAdapters(CAdapterGenerator.kt:173)
	at org.jetbrains.kotlin.backend.konan.ExportedElementScope.generateCAdapters(CAdapterGenerator.kt:173)
	at org.jetbrains.kotlin.backend.konan.ExportedElementScope.generateCAdapters(CAdapterGenerator.kt:173)
	at org.jetbrains.kotlin.backend.konan.ExportedElementScope.generateCAdapters(CAdapterGenerator.kt:173)
	at org.jetbrains.kotlin.backend.konan.ExportedElementScope.generateCAdapters(CAdapterGenerator.kt:173)
	at org.jetbrains.kotlin.backend.konan.CAdapterGenerator.generateBindings(CAdapterGenerator.kt:714)
	at org.jetbrains.kotlin.backend.konan.CAdapterGenerator.generateBindings(CAdapterGenerator.kt:690)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.appendCAdapters(IrToBitcode.kt:280)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:309)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitModuleFragment(IrElementVisitorVoid.kt:28)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:181)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:181)
	at org.jetbrains.kotlin.ir.declarations.impl.IrModuleFragmentImpl.accept(IrModuleFragmentImpl.kt:41)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptVoid(IrElementVisitorVoid.kt:258)
	at org.jetbrains.kotlin.backend.konan.llvm.BitcodePhasesKt$codegenPhase$1.invoke(BitcodePhases.kt:205)
	at org.jetbrains.kotlin.backend.konan.llvm.BitcodePhasesKt$codegenPhase$1.invoke(BitcodePhases.kt)
	at org.jetbrains.kotlin.backend.konan.KonanLoweringPhasesKt$makeKonanModuleOpPhase$1.invoke(KonanLoweringPhases.kt:57)
	at org.jetbrains.kotlin.backend.konan.KonanLoweringPhasesKt$makeKonanModuleOpPhase$1.invoke(KonanLoweringPhases.kt:55)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:128)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:24)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:127)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:105)
	at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:128)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:24)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:127)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:105)
	at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:128)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:24)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:127)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:105)
	at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:128)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:24)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:127)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:105)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:42)
	at org.jetbrains.kotlin.backend.konan.KonanDriverKt.runTopLevelPhases(KonanDriver.kt:27)
	at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:79)
	at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:34)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:84)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:42)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:104)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:82)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:50)
	at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit(CLITool.kt:215)
	at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMain(CLITool.kt:207)
	at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:223)
	at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:220)
	at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:26)
	at org.jetbrains.kotlin.konan.util.UtilKt.profile(Util.kt:21)
	at org.jetbrains.kotlin.cli.bc.K2Native$Companion.main(K2Native.kt:222)
	at org.jetbrains.kotlin.cli.bc.K2NativeKt.main(K2Native.kt:304)
	at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:16)


> Task :linkDebugSharedMingw FAILED
```

**To Reproduce**
```kotlin
interface LowercaseString {
    val lowercase: String
}

@Serializable
data class MyData(val maybeCapializedString: String) : LowercaseString {
    constructor() : this("")

    override fun toString() = maybeCapializedString.uncapitalize()

    override val lowercase: String
        get() = maybeCapializedString.uncapitalize()

    init { freeze() }

    @Serializer(forClass = MyData::class)
    companion object : KSerializer<MyData> {
        override val descriptor: SerialDescriptor = StringDescriptor.withName("MyDataDescriptor")

        override fun serialize(encoder: Encoder, obj: MyData) {
            encoder.encodeString(obj.lowercase)
        }

        override fun deserialize(decoder: Decoder): MyData {
            return MyData(decoder.decodeString())
        }
    }
}
```

**Expected behavior**
Builds (works fine for Apple and JVM targets)

**Environment**
 - Kotlin version: 1.3.50
 - Library version: 0.12.0
 - Kotlin platforms: Native - mingw
 - Gradle version: 5.6.1
 - IDE version (if bug is related to the IDE) [e.g. IntellijIDEA 2019.1, Android Studio 3.4]
 - Other relevant context [e.g. OS version, JRE version, ... ]

**Comment by benasher44** [comment]
I tried 0.13.0 and got the same result.

**Comment by benasher44** [comment]
This is fixed in Kotlin 1.3.60 / serialization 0.14.0 for me
